zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py

```diff
@@ -0,0 +1,119 @@
+# Generated by Django 4.2.3 on 2023-08-16 14:16
+
+import bitfield.models
+from django.contrib.postgres.operations import AddIndexConcurrently, RemoveIndexConcurrently
+from django.db import migrations, models
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Q
+
+
+def clear_set_bit_topic_wildcard_and_group_mentioned(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    UserMessage = apps.get_model("zerver", "UserMessage")
+
+    ums = UserMessage.objects.filter(
+        Q(flags__andnz=UserMessage.flags.mentioned)
+        | Q(flags__andnz=UserMessage.flags.stream_wildcard_mentioned)
+        | Q(flags__andnz=UserMessage.flags.topic_wildcard_mentioned)
+        | Q(flags__andnz=UserMessage.flags.group_mentioned)
+    ).values_list("user_profile_id", "message_id")
+
+    message_ids = set()
+    user_profile_ids = set()
+    for user_profile_id, message_id in ums:
+        message_ids.add(message_id)
+        user_profile_ids.add(user_profile_id)
+
+    user_messages = UserMessage.objects.filter(
+        user_profile_id__in=user_profile_ids, message_id__in=message_ids
```

Comment:
This sort of `EXPLAIN ANALYZE` check isn't very helpful in the development environment, just because the data sizes being small can warp what index ends up being used.

I think what we want to do for this migration is something following some of the patterns in  `zerver/migrations/0239_usermessage_copy_id_to_bigint_id.py`. But roughly, we're going to want to have a loop processing batches of say 1K-10K `UserMessage` rows that match our query, ideally doing the work all in SQL, with progress indication print statements, and definitely avoid sending lists of individual message/user IDs to the database in each request, if we can.

You may want to do a pairing session with @mateuszmandera to work on this, as he has a lot of experience writing this sort of migration on our biggest tables.

Reply:
Yeah.  Regardless of if that query is indexed, the fetch of the initial `ums` variable for Zulip Cloud is going to be too large to comfortably fit in memory.  In addition to being slow because it's done in Python, this also has the difficulty of exposing us to race conditions between when we fetch `user_messages` and when we update using `bulk_update`.  Any solution here is going to need to use `update` and run the logic entirely in PostgreSQL, in batches.

Reply:
0486 has the related migration